File: backend/app/api/v1/endpoints/contents.py
"""Endpoints de `Content`: a entidade central do produto."""

from __future__ import annotations

import uuid
from datetime import date

from fastapi import APIRouter, BackgroundTasks, Query, status

from app.core.deps import CurrentBusiness, CurrentUser, DbSession
from app.models.content import Content
from app.models.enums import ContentFormat, ContentObjective, ContentStatus, RegenerationScope
from app.schemas.asset import AssetRead
from app.schemas.common import MessageResponse, Page
from app.schemas.content import (
    ContentActionResponse,
    ContentAssetLinkRequest,
    ContentAssetRead,
    ContentChangeFormatRequest,
    ContentDuplicateRequest,
    ContentFromIdeaRequest,
    ContentGenerateRequest,
    ContentManualCreate,
    ContentRead,
    ContentRegenerateRequest,
    ContentScheduleRequest,
    ContentStatusUpdate,
    ContentSummary,
    ContentUpdate,
    ContentVersionRead,
    CreationQuestion,
)
from app.schemas.job import JobAccepted
from app.services.ai_service import AIService
from app.services.asset_service import AssetService
from app.services.content_service import ContentService
from app.services.creation_questions import list_creation_questions, resolve_creation_item
from app.workers.dispatcher import schedule_ai_job

router = APIRouter(prefix="/contents", tags=["contents"])


def _to_read(content: Content, assets: AssetService) -> ContentRead:
    data = ContentRead.model_validate(content)
    linked = [
        ContentAssetRead(
            asset=AssetRead.model_validate(link.asset).model_copy(
                update={"url": assets.signed_url(link.asset)}
            ),
            role=link.role,
            position=link.position,
        )
        for link in content.asset_links
        if link.asset is not None
    ]
    presenter = (content.generation_context or {}).get("presenter") or {}
    return data.model_copy(
        update={
            "assets": linked,
            "presenter_name": presenter.get("display_name"),
            "allowed_transitions": ContentService.allowed_transitions(content),
        }
    )


# ------------------------------------------------------------------- criacao --


@router.post(
    "/from-idea",
    response_model=JobAccepted,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Etapa 3 e 4: transforma uma ideia escolhida em Content",
)
async def create_from_idea(
    payload: ContentFromIdeaRequest,
    business: CurrentBusiness,
    user: CurrentUser,
    session: DbSession,
    background: BackgroundTasks,
) -> JobAccepted:
    job = await AIService(session, business=business, user_id=user.id).request_production(payload)
    schedule_ai_job(job.id, background)
    return JobAccepted(job_id=job.id, status=job.status, kind=job.kind)


@router.get(
    "/generate/questions",
    response_model=list[CreationQuestion],
    summary="Perguntas que faltam antes de gerar (no maximo 3)",
)
async def list_generate_questions(
    business: CurrentBusiness,
    session: DbSession,
    objective: ContentObjective = Query(...),
    product_id: uuid.UUID | None = Query(default=None),
    service_id: uuid.UUID | None = Query(default=None),
) -> list[CreationQuestion]:
    product, service = await resolve_creation_item(
        session,
        business.id,
        objective=objective,
        product_id=product_id,
        service_id=service_id,
    )
    return await list_creation_questions(
        session, business, objective=objective, product=product, service=service
    )


@router.post(
    "/generate",
    response_model=JobAccepted,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Cria conteudo em um passo (ideia + producao no mesmo job)",
)
async def generate_content(
    payload: ContentGenerateRequest,
    business: CurrentBusiness,
    user: CurrentUser,
    session: DbSession,
    background: BackgroundTasks,
) -> JobAccepted:
    job = await AIService(session, business=business, user_id=user.id).request_content_creation(
        payload
    )
    schedule_ai_job(job.id, background)
    return JobAccepted(job_id=job.id, status=job.status, kind=job.kind)


@router.post("", response_model=ContentRead, status_code=status.HTTP_201_CREATED)
async def create_manual(
    payload: ContentManualCreate, business: CurrentBusiness, session: DbSession
) -> ContentRead:
    """Cria um conteudo do zero, sem passar pela IA."""
    service = ContentService(session)
    content = await service.create_manual(business.id, payload)
    return _to_read(await service.get(business.id, content.id), AssetService(session))


# ------------------------------------------------------------------- leitura --


@router.get("", response_model=Page[ContentSummary])
async def list_contents(
    business: CurrentBusiness,
    session: DbSession,
    content_status: list[ContentStatus] | None = Query(default=None, alias="status"),
    content_format: list[ContentFormat] | None = Query(default=None, alias="format"),
    category: str | None = Query(default=None),
    search: str | None = Query(default=None, max_length=120),
    planned_from: date | None = Query(default=None),
    planned_to: date | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
) -> Page[ContentSummary]:
    items, total = await ContentService(session).list(
        business.id,
        statuses=content_status or (),
        formats=content_format or (),
        category=category,
        search=search,
        planned_from=planned_from,
        planned_to=planned_to,
        limit=limit,
        offset=offset,
    )
    return Page[ContentSummary](
        items=[ContentSummary.model_validate(item) for item in items],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/{content_id}", response_model=ContentRead)
async def get_content(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> ContentRead:
    content = await ContentService(session).get(business.id, content_id)
    return _to_read(content, AssetService(session))


@router.get("/{content_id}/versions", response_model=list[ContentVersionRead])
async def list_versions(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> list[ContentVersionRead]:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    versions = await service.list_versions(content)
    return [ContentVersionRead.model_validate(version) for version in versions]


# -------------------------------------------------------------------- edicao --


@router.patch("/{content_id}", response_model=ContentRead)
async def update_content(
    content_id: uuid.UUID,
    payload: ContentUpdate,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.update(content, payload)
    return _to_read(updated, AssetService(session))


@router.post(
    "/{content_id}/regenerate",
    response_model=JobAccepted,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Regenera o conteudo inteiro ou apenas uma parte",
)
async def regenerate_content(
    content_id: uuid.UUID,
    payload: ContentRegenerateRequest,
    business: CurrentBusiness,
    user: CurrentUser,
    session: DbSession,
    background: BackgroundTasks,
) -> JobAccepted:
    job = await AIService(session, business=business, user_id=user.id).request_regeneration(
        content_id, payload
    )
    schedule_ai_job(job.id, background)
    return JobAccepted(job_id=job.id, status=job.status, kind=job.kind)


@router.post("/{content_id}/change-format", response_model=ContentActionResponse)
async def change_format(
    content_id: uuid.UUID,
    payload: ContentChangeFormatRequest,
    business: CurrentBusiness,
    user: CurrentUser,
    session: DbSession,
    background: BackgroundTasks,
) -> ContentActionResponse:
    """Troca o formato e, por padrao, reescreve o conteudo no novo formato.

    A troca em si e sincrona (o payload antigo nao vale no formato novo); a
    reescrita vai para a fila porque depende da IA.
    """
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    content = await service.change_format(content, payload.format, reset_payload=True)

    job_id = None
    if payload.regenerate:
        job = await AIService(session, business=business, user_id=user.id).request_regeneration(
            content.id,
            ContentRegenerateRequest(
                scope=RegenerationScope.FULL, instruction=payload.instruction
            ),
        )
        schedule_ai_job(job.id, background)
        job_id = job.id

    refreshed = await service.get(business.id, content.id)
    return ContentActionResponse(
        content=_to_read(refreshed, AssetService(session)), job_id=job_id
    )


@router.post("/{content_id}/duplicate", response_model=ContentRead, status_code=status.HTTP_201_CREATED)
async def duplicate_content(
    content_id: uuid.UUID,
    payload: ContentDuplicateRequest,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    copy = await service.duplicate(content, title=payload.title)
    return _to_read(await service.get(business.id, copy.id), AssetService(session))


@router.post("/{content_id}/versions/{version}/restore", response_model=ContentRead)
async def restore_version(
    content_id: uuid.UUID,
    version: int,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    restored = await service.restore_version(content, version)
    return _to_read(restored, AssetService(session))


# ---------------------------------------------------------------- transicoes --


@router.post("/{content_id}/status", response_model=ContentRead)
async def change_status(
    content_id: uuid.UUID,
    payload: ContentStatusUpdate,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.change_status(
        content, payload.status, planned_date=payload.planned_date, reason=payload.reason
    )
    return _to_read(updated, AssetService(session))


@router.post("/{content_id}/approve", response_model=ContentRead)
async def approve_content(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.change_status(content, ContentStatus.APPROVED)
    return _to_read(updated, AssetService(session))


@router.post("/{content_id}/reject", response_model=ContentRead)
async def reject_content(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.change_status(content, ContentStatus.REJECTED)
    return _to_read(updated, AssetService(session))


@router.post("/{content_id}/archive", response_model=ContentRead)
async def archive_content(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.change_status(content, ContentStatus.ARCHIVED)
    return _to_read(updated, AssetService(session))


@router.post("/{content_id}/schedule", response_model=ContentRead)
async def schedule_content(
    content_id: uuid.UUID,
    payload: ContentScheduleRequest,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    """Define a data planejada.

    A publicacao automatica no Instagram ainda nao existe: aqui o conteudo e
    organizado no calendario e, se estiver aprovado, passa a SCHEDULED.
    """
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.schedule(content, payload.planned_date)
    return _to_read(updated, AssetService(session))


@router.delete("/{content_id}", response_model=MessageResponse)
async def delete_content(
    content_id: uuid.UUID, business: CurrentBusiness, session: DbSession
) -> MessageResponse:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    await service.delete(content)
    return MessageResponse(message="Conteudo excluido.")


# -------------------------------------------------------------------- assets --


@router.post("/{content_id}/assets", response_model=ContentRead)
async def link_asset(
    content_id: uuid.UUID,
    payload: ContentAssetLinkRequest,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.link_asset(
        content, payload.asset_id, role=payload.role, position=payload.position
    )
    return _to_read(updated, AssetService(session))


@router.delete("/{content_id}/assets/{asset_id}", response_model=ContentRead)
async def unlink_asset(
    content_id: uuid.UUID,
    asset_id: uuid.UUID,
    business: CurrentBusiness,
    session: DbSession,
) -> ContentRead:
    service = ContentService(session)
    content = await service.get(business.id, content_id)
    updated = await service.unlink_asset(content, asset_id)
    return _to_read(updated, AssetService(session))

File: frontend/src/lib/api/contents.ts
import { api } from "./client";
import type {
  ContentActionResponse,
  ContentAssetRole,
  ContentFormat,
  ContentObjective,
  ContentRead,
  ContentStatus,
  ContentSummary,
  ContentVersionRead,
  CreationQuestion,
  JobAccepted,
  Page,
  RegenerationScope,
  VisualTone,
} from "./types";

export const contentsApi = {
  generate: (payload: {
    product_id?: string | null;
    service_id?: string | null;
    objective: ContentObjective;
    format?: ContentFormat | null;
    answers?: Record<string, string>;
    planned_date?: string | null;
    visual_tone?: VisualTone;
  }) => api.post<JobAccepted>("/contents/generate", payload),

  generateQuestions: (params: {
    objective: ContentObjective;
    product_id?: string | null;
    service_id?: string | null;
  }) =>
    api.get<CreationQuestion[]>("/contents/generate/questions", {
      query: {
        objective: params.objective,
        product_id: params.product_id,
        service_id: params.service_id,
      },
    }),

  createFromIdea: (payload: {
    idea_id: string;
    format?: ContentFormat | null;
    instruction?: string | null;
    planned_date?: string | null;
  }) => api.post<JobAccepted>("/contents/from-idea", payload),

  createManual: (payload: {
    title: string;
    format: ContentFormat;
    category?: string | null;
    concept?: string | null;
    objective?: string | null;
    caption?: string | null;
    cta?: string | null;
    hashtags?: string[];
    planned_date?: string | null;
  }) => api.post<ContentRead>("/contents", payload),

  list: (filters?: {
    status?: ContentStatus[];
    format?: ContentFormat[];
    category?: string;
    search?: string;
    plannedFrom?: string;
    plannedTo?: string;
    limit?: number;
    offset?: number;
  }) =>
    api.get<Page<ContentSummary>>("/contents", {
      query: {
        status: filters?.status,
        format: filters?.format,
        category: filters?.category,
        search: filters?.search,
        planned_from: filters?.plannedFrom,
        planned_to: filters?.plannedTo,
        limit: filters?.limit,
        offset: filters?.offset,
      },
    }),

  get: (id: string) => api.get<ContentRead>(`/contents/${id}`),

  versions: (id: string) => api.get<ContentVersionRead[]>(`/contents/${id}/versions`),

  update: (
    id: string,
    payload: Partial<{
      title: string;
      concept: string | null;
      objective: string | null;
      category: string | null;
      caption: string | null;
      cta: string | null;
      hashtags: string[];
      planned_date: string | null;
      payload: Record<string, unknown>;
      change_reason: string | null;
    }>
  ) => api.patch<ContentRead>(`/contents/${id}`, payload),

  regenerate: (
    id: string,
    payload: { scope: RegenerationScope; instruction?: string | null }
  ) => api.post<JobAccepted>(`/contents/${id}/regenerate`, payload),

  changeFormat: (
    id: string,
    payload: { format: ContentFormat; regenerate?: boolean; instruction?: string | null }
  ) => api.post<ContentActionResponse>(`/contents/${id}/change-format`, payload),

  duplicate: (id: string, title?: string | null) =>
    api.post<ContentRead>(`/contents/${id}/duplicate`, { title }),

  restoreVersion: (id: string, version: number) =>
    api.post<ContentRead>(`/contents/${id}/versions/${version}/restore`),

  changeStatus: (
    id: string,
    payload: { status: ContentStatus; planned_date?: string | null; reason?: string | null }
  ) => api.post<ContentRead>(`/contents/${id}/status`, payload),

  approve: (id: string) => api.post<ContentRead>(`/contents/${id}/approve`),
  reject: (id: string) => api.post<ContentRead>(`/contents/${id}/reject`),
  archive: (id: string) => api.post<ContentRead>(`/contents/${id}/archive`),

  schedule: (id: string, plannedDate: string) =>
    api.post<ContentRead>(`/contents/${id}/schedule`, { planned_date: plannedDate }),

  remove: (id: string) => api.delete<{ message: string }>(`/contents/${id}`),

  linkAsset: (id: string, payload: { asset_id: string; role?: ContentAssetRole; position?: number }) =>
    api.post<ContentRead>(`/contents/${id}/assets`, payload),

  unlinkAsset: (id: string, assetId: string) =>
    api.delete<ContentRead>(`/contents/${id}/assets/${assetId}`),
};
